File: publishing/domain/collection/publication.rs
use common::result::Result;

use crate::domain::author::AuthorId;
use crate::domain::category::CategoryId;
use crate::domain::publication::PublicationId;

#[derive(Debug, Clone)]
pub struct Publication {
    id: PublicationId,
    author_id: AuthorId,
    name: String,
    synopsis: String,
    category_id: CategoryId,
    tags: Vec<String>,
    cover: String,
    pages: u32,
}

impl Publication {
    pub fn new<S: Into<String>>(
        id: PublicationId,
        author_id: AuthorId,
        name: S,
        synopsis: S,
        category_id: CategoryId,
        tags: Vec<String>,
        cover: S,
        pages: u32,
    ) -> Result<Self> {
        Ok(Publication {
            id,
            author_id,
            name: name.into(),
            synopsis: synopsis.into(),
            category_id,
            tags,
            cover: cover.into(),
            pages,
        })
    }

    pub fn id(&self) -> &PublicationId {
        &self.id
    }

    pub fn author_id(&self) -> &AuthorId {
        &self.author_id
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn synopsis(&self) -> &str {
        &self.synopsis
    }

    pub fn category_id(&self) -> &CategoryId {
        &self.category_id
    }

    pub fn tags(&self) -> &[String] {
        &self.tags
    }

    pub fn cover(&self) -> &str {
        &self.cover
    }
}

File: web/src/app/domain/models/publication.ts
import { IAuthor } from './author';
import { ICategory } from './category';
import { IStatistics } from './statistics';

export interface IImage {
  url: string;
}

export interface IPage {
  number: number;
  images: IImage;
}

export interface IStatus {
  status: string;
  changed_by?: string;
  comment?: string;
}

export interface IPublication {
  id: string;
  author_id?: string;
  author?: IAuthor;
  name: string;
  synopsis: string;
  category_id?: string;
  category?: ICategory;
  tags: string[];
  cover: string;
  statistics?: IStatistics;
  status: IStatus;
  pages?: IPage[];
  created_at: string;
  updated_at?: string;
}
